controller/internal/supervisor/iptablesctrl/acls_windows.go

```diff
@@ -0,0 +1,218 @@
+	// delete individual rules. This is because Windows uses non-PU-specific chains.
+
+	// delete ACLs rules for Windows
+	appACLIPset := i.getACLIPSets(containerInfo.Policy.ApplicationACLs())
+	netACLIPset := i.getACLIPSets(containerInfo.Policy.NetworkACLs())
+
+	if err := i.deleteExternalACLs(cfg, cfg.AppChain, cfg.NetChain, appACLIPset, true); err != nil {
+		zap.L().Warn("Failed to delete ACL rules for app chain", zap.Error(err))
+	}
+
+	if err := i.deleteExternalACLs(cfg, cfg.NetChain, cfg.AppChain, netACLIPset, false); err != nil {
+		zap.L().Warn("Failed to delete ACL rules for net chain", zap.Error(err))
+	}
+
+	// delete deny-all rules for Windows
+	isHostPU := extractors.IsHostPU(containerInfo.Runtime, i.mode)
+	return i.processRulesFromList(i.trapRules(cfg, isHostPU, [][]string{}, [][]string{}), "Delete")
+}
+
+func (i *iptables) deleteExternalACLs(cfg *ACLInfo, chain string, reverseChain string, rules []aclIPset, isAppAcls bool) error {
+
+	_, rules = extractProtocolAnyRules(rules)
+
+	rulesBucket := i.sortACLsInBuckets(cfg, chain, reverseChain, rules, isAppAcls)
+
+	aclRules, err := extractACLsFromTemplate(rulesBucket)
+	if err != nil {
+		return fmt.Errorf("unable to extract rules from template: %s", err)
+	}
+
+	aclRules = transformACLRules(aclRules, cfg, rulesBucket, isAppAcls)
+
+	if err := i.processRulesFromList(aclRules, "Delete"); err != nil {
+		return fmt.Errorf("unable to delete rules - mode :%s %v", err, isAppAcls)
+	}
+
+	return nil
+}
+
+// removeGlobalHooksPre for Windows does nothing
+func (i *iptables) removeGlobalHooksPre() {
+}
+
+// try to merge two acl rules (one log and one accept/drop) into one for Windows
+func makeTerminatingRuleFromPair(aclRule1, aclRule2 []string) *winipt.WindowsRuleSpec {
+
+	if aclRule1 == nil || aclRule2 == nil {
+		return nil
+	}
+	winRuleSpec1, err := winipt.ParseRuleSpec(aclRule1[2:]...)
+	if err != nil {
+		return nil
+	}
+	winRuleSpec2, err := winipt.ParseRuleSpec(aclRule2[2:]...)
+	if err != nil {
+		return nil
+	}
+
+	// save action/log properties
+	action := winRuleSpec1.Action
+	logPrefix := winRuleSpec2.LogPrefix
+	groupId := winRuleSpec2.GroupId
+	if action == 0 {
+		action = winRuleSpec2.Action
+		if action == 0 {
+			return nil
+		}
+		logPrefix = winRuleSpec1.LogPrefix
+		groupId = winRuleSpec1.GroupId
+		if !winRuleSpec1.Log {
+			return nil
+		}
+	} else if !winRuleSpec2.Log {
+		return nil
+	}
+
+	// if one is nflog and one is another action, and they are otherwise equal, then combine into one rule
+	winRuleSpec1.Log = false
+	winRuleSpec1.LogPrefix = ""
+	winRuleSpec1.GroupId = 0
+	winRuleSpec1.Action = 0
+	winRuleSpec2.Log = false
+	winRuleSpec2.LogPrefix = ""
+	winRuleSpec2.GroupId = 0
+	winRuleSpec2.Action = 0
+	if reflect.DeepEqual(winRuleSpec1, winRuleSpec2) {
+		winRuleSpec1.Log = true
+		winRuleSpec1.LogPrefix = logPrefix
+		winRuleSpec1.GroupId = groupId
+		winRuleSpec1.Action = action
+		return winRuleSpec1
+	}
+	return nil
```

Comment:
That gives me a little pain, because I am not even sure it always works and it is actually very slow. If we can find a way to avoid using reflect.DeepEqual I would prefer it.

Reply:
fixed by right my own Equal comparison